extensions/aws/s3/S3ClientRequestSender.h

```diff
@@ -0,0 +1,47 @@
+/**
+ * @file S3Wrapper.h
+ * S3Wrapper class declaration
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "S3RequestSender.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace aws {
+namespace s3 {
+
+class S3ClientRequestSender : public S3RequestSender {
+ public:
+  minifi::utils::optional<Aws::S3::Model::PutObjectResult> sendPutObjectRequest(const Aws::S3::Model::PutObjectRequest& request) override;
+  bool sendDeleteObjectRequest(const Aws::S3::Model::DeleteObjectRequest& request) override;
+  minifi::utils::optional<Aws::S3::Model::GetObjectResult> sendGetObjectRequest(const Aws::S3::Model::GetObjectRequest& request) override;
+  minifi::utils::optional<Aws::S3::Model::ListObjectsV2Result> sendListObjectsRequest(const Aws::S3::Model::ListObjectsV2Request& request) override;
+  minifi::utils::optional<Aws::S3::Model::ListObjectVersionsResult> sendListVersionsRequest(const Aws::S3::Model::ListObjectVersionsRequest& request) override;
+  minifi::utils::optional<Aws::S3::Model::GetObjectTaggingResult> sendGetObjectTaggingRequest(const Aws::S3::Model::GetObjectTaggingRequest& request) override;
+  minifi::utils::optional<Aws::S3::Model::HeadObjectResult> sendHeadObjectRequest(const Aws::S3::Model::HeadObjectRequest& request) override;
+};
```

Comment:
The ListS3 processor is supposed to be the last S3 processor for the time being, I'm not sure if we have any additional S3 functionality in the future. Do you have any particular use case in mind for the async API? In that case I suppose we should have the same async API in the S3Wrapper as well, to be able to get the processor-desired result in an async way. In my opinion we should only invest in implementing and testing an async API when we have a use case for it, I'm not sure if it would be worth right now.

Reply:
Async operations generally scale better, that's why I was raising this. The functionality and use cases are the same, but it would probably result in 1-2 orders of magnitude better performance. Now we have to wait for a request to finish before we can proceed. We could pipeline multiple requests with an async API.

Reply:
I think the async API could help us in a way that currently we always wait for the request to finish because of the `s3_wrapper_mutex_` which helps to avoid overwriting client configuration of the S3 client in concurrent `onTrigger` calls. Because of the mutex only 1 request can run at a time, and you are right that making the client async, waiting for the result could be done afterwards. Another solution I was thinking about is making the S3Wrapper stateless and passing all the configuration parameters to every single call, making it thread safe to call `onTrigger` function parallel which would create separate clients with separate configurations. What do you think?

Reply:
Regarding S3 client, if it's cheap to create, then I would go for solution 2, otherwise solution 1. But you're more familiar with the ecosystem, so I trust your judgement.

In any case, an easy future-proofing (pun intended) could be to wrap the current solution in `async(deferred, ...)` and convert to actual async later. After giving this a second thought, async conversion would probably require other changes as well, so this may not be a good idea after all.

It's your call whether you see any value in any of the ideas raised. I'm fine with no changes as well.

Reply:
I think it's a good and idea to improve upon this and it should be investigated further or rather experimented with the available solutions. As it impacts all the S3 API and the callbacks in each S3 processor and test and therefore all S3 processors themselves, I would move this to a followup PR so I created a separate Jira ticket for this: https://issues.apache.org/jira/browse/MINIFICPP-1476